chore: update @appium/fake-driver to use @appium/base-driver

boneskull opened the pull request:
BREAKING CHANGE: `fake-driver` now depends upon `@appium/base-driver@8.x`

## `@appium/fake-driver`

- need to use w3c capabilities only
- switch from `wd` to `wdio` in tests
- many, many fixes to support these

## `@appium/base-driver`

- fixed a dead URL in a comment
- updated the "logging" tests to manually supply w3c capabilities.  `createSession()` does it for you, but `executeCommand('createSession')` does not.
- display the name of the driver under test when executing base driver's test suite with other drivers
- add missing `releaseActions` endpoint
 
Ref: #15436

boneskull committed:
chore: update @appium/fake-driver to use @appium/base-driver

BREAKING CHANGE: `fake-driver` now depends upon `@appium/base-driver@8.x`

## `@appium/fake-driver`

- need to use w3c capabilities only
- fix: find `app.xml` fixture properly when running tests via `mocha --require=@babel/register`

## `@appium/base-driver`

- fixed a dead URL in a comment
- updated the "logging" tests to manually supply w3c capabilities.  `createSession()` does it for you, but `executeCommand('createSession')` does not.
- display the name of the driver under test when executing base driver's test suite with other drivers

Ref: #15425

boneskull commented on packages/fake-driver/lib/logger.js:
```diff
@@ -1,4 +1,4 @@
-import { logger } from 'appium-support';
+import { logger } from '@appium/support';
```
I am not entirely sure why this wasn't caught before.

jlipps replied:
probably some other package still had appium-support as a dep?

boneskull commented on packages/fake-driver/test/helpers.js:
```diff
@@ -1,17 +1,21 @@
+// XXX: what dir are we in? it depends on how we're running this test. the .xml file doesn't move, but we do!
+// we may be running with @babel/register; otherwise, we're probably in `build/test/` which _does not_ contain the xml file.
+const appFixturePath = path.join(__dirname, 'fixtures', 'app.xml');
+const TEST_APP = existsSync(appFixturePath) ? appFixturePath : path.join(__dirname, '..', '..', 'test', 'fixtures', 'app.xml');
```
two potential fixes here

1. stop using babel 😄
2. copy the xml file into `build` during transpilation

mykola-mokhnach replied:
option #2 looks reasonable to be

jlipps replied:
yeah this bites us in a few different places, where we have to unreasonably path out to the package root and then down into the build dir. if there's an easy way to copy fixtures during transpilation, great!

jlipps replied:
should we create an issue to track this task?

KazuCocoa commented on packages/fake-driver/package.json:
```diff
@@ -30,9 +30,9 @@
     "screen.png"
   ],
   "dependencies": {
+    "@appium/base-driver": "^8.0.0-beta.6",
```
👍

KazuCocoa reviewed (approved):
👍 
lint? :)

mykola-mokhnach commented on packages/base-driver/test/basedriver/driver-tests.js:
```diff
@@ -75,12 +78,12 @@ function baseDriverUnitTests (DriverClass, defaultCaps = {}) {
+        id: d.sessionId
+      });
+      sessions[0].capabilities.should.include({
+        'deviceName': 'Commodore 64',
```
I assume there is no need to quote key names

boneskull replied:
no, but we don't auto-format the code and eslint doesn't seem to have a rule enabled about it, so ... we should address that

jlipps replied:
i thought we had an eslint rule. i guess not. we should add a rule

jlipps replied:
even if there's not a rule, let's unquote these key names as part of this review

mykola-mokhnach commented on packages/fake-driver/lib/fake-app.js:
```diff
@@ -1,13 +1,11 @@
-import B from 'bluebird';
-import fs from 'fs';
+import {readFile} from 'fs/promises';
```
readFile with promises could be imported from fs submodule of appium-support

boneskull replied:
I'm curious: why promisify the `fs` api with bluebird?  performance reasons?

jlipps replied:
Before promise versions of the node apis were shipped with Node, we put promisified versions in appium-support since we used them everywhere. I'm not sure why we would keep using them in appium-support though now, unless there are functional differences (@mykola-mokhnach ?)

boneskull replied:
(I've pushed this change)

mykola-mokhnach replied:
> Before promise versions of the node apis were shipped with Node, we put promisified versions in appium-support since we used them everywhere. I'm not sure why we would keep using them in appium-support though now, unless there are functional differences (@mykola-mokhnach ?)

Exactly. As for appium-support we could now replace bluebird usage with async file operations natively provided by node. I don't see any point in changing method signatures because it would require too many changes in depending modules

mykola-mokhnach commented:
Tests are still not quite happy though

jlipps commented on packages/fake-driver/lib/fake-app.js:
```diff
@@ -1,13 +1,11 @@
-import B from 'bluebird';
-import fs from 'fs';
+import {readFile} from 'fs/promises';
```
i don't know if we can have an eslint rule about this, but we differentiate named imports from object destructuring in our whitespace:

```
import { foo } from 'bar';
export { thing1, thing2 };
```
vs
```
const {foo} = bar;
const thing = {x: 1, y: 2};
```

boneskull replied:
eslint is obviously not configured to care about this...  I do pretty well with an eslint + prettier combo and would be gung-ho about adding prettier here, if you are interested.

jlipps replied:
I'm up for that. Are you saying it could help us make this distinction?

boneskull replied:
I tried it, actually, and prettier makes no distinction.  you may have a tough time making this work.

boneskull committed:
fix(fake-driver): add naive css selector support

boneskull committed:
fix(fake-driver): helpers now use wdio instead of wd

- exported more constants as base-driver helpers do
- `deleteSession()` now expects a wdio `browser` (`driver`) object; removed closure

boneskull committed:
fix(fake-driver): find-element-tests and driver-e2e-specs members passing

- alert-tests, context-tests, element-interaction-tests, general-tests all failing still

boneskull committed:
fix(fake-driver): update interaction tests

replace unsupported `getLocation` with `getElementRect`

boneskull committed:
fix(fake-driver): update alert tests

boneskull committed:
fix(fake-driver): udpate context tests

boneskull committed:
fix(fake-driver): update unit tests

boneskull committed:
fix(base-driver): implement missing route for releaseActions

is the call to `this.proxyReleaseActions` even necessary?

boneskull committed:
fix(fake-driver): update generic actions tests

also add a stub implementation of `releaseActions`, which wdio calls

boneskull committed:
chore(fake-driver): remove redundant tests from find-element-tests

covered by other strategies

boneskull committed:
fix(fake-driver): update generic actions tests

also fix path to screenshot (`screen.png`) when running tests outside of `build`

boneskull commented:
test failure is expected since I added a route... unfortunately that seems to prevent the rest of the suite from running.

jlipps commented:
The solution is to update that failing test with the (now) correct hash: `3351a966`

boneskull committed:
chore(fake-driver): use fs module from @appium/support

boneskull committed:
chore(base-driver): update hash in protocol consistency test

boneskull committed:
fix(base-driver): update e2e test fn to prefer provided address/port over default

jlipps commented on packages/fake-driver/lib/fake-app.js:
```diff
@@ -1,14 +1,13 @@
-import B from 'bluebird';
-import fs from 'fs';
+import {fs} from '@appium/support';
```
even though there's not a rule, let's stick with the `{ <name> }` spacing style if we notice it.

jlipps commented on packages/fake-driver/test/general-tests.js:
```diff
@@ -1,7 +1,7 @@
 import chaiAsPromised from 'chai-as-promised';
-import { initSession, deleteSession, DEFAULT_CAPS } from './helpers';
-import { W3CActions } from 'wd';
+import chaiWebdriverIOAsync from 'chai-webdriverio-async';
```
o_O there's a package for this??

boneskull replied:
yeah.  async webdriver isn't the most fun imo... this makes it a little easier. sometimes.

jlipps reviewed (changes requested):
couple style nits, otherwise this is ready to go!

boneskull committed:
chore: formatting

boneskull commented:
@jlipps Created #15471

jlipps commented:
thanks so much @boneskull! this was incredible, and no doubt incredibly tedious, work.
